clippy_lints/src/eta_reduction.rs

```diff
@@ -174,6 +190,42 @@ fn check_closure(cx: &LateContext<'_>, expr: &Expr<'_>) {
     }
 }
 
+fn used_after_expr<'tcx>(mut visitor: LocalUsedVisitor<'tcx>, block: &Block<'tcx>, expr: &Expr<'_>) -> bool {
```

Comment:
This impl looks wrong/unnecessary to me. Staring at this for a while, I think I figured out why:

1. You use spans to check where an expression is. This will probably fail in some cases, for example, when macros are involved. 
2. The loop detection seems too complicated to me. This is the HIR, so you shouldn't have to differentiate between loop kinds.
3. We have a function  `clippy_utils::get_enclosing_loop_or_closure` you can just use to check if the closure is called in a context where `&mut` is necessary.
4. A "used after `expr`" sounds like something we're already doing in other places. Looking for it, I only found lint specific implementations though. But I think that this is a good opportunity to implement a visitor for this in `clippy_utils::usage`. With a visitor you can check if you already passed an expression like this: https://github.com/rust-lang/rust-clippy/blob/64d74df19c582a10ad41ddce496528e597b06765/clippy_lints/src/loops/utils.rs#L175-L176 This should be implemented as a visitor.

Reply:
Thanks for your help! Can you point me towards the other lints that implement a "used after `expr`" so that I can switch to the visitor there, or should that be a separate PR?

Reply:
If at all this should be a separate PR. This is mostly done in other visitors that are pretty specific to the lints they're used in, so I'm not sure if it's worth splitting that out.